File: src/main/java/serialize/Deserialize.java
package serialize;

import javafx.scene.layout.Pane;
import javafx.scene.shape.Line;
import optics.PreciseLine;
import optics.light.Ray;
import optics.objects.Mirror;
import optics.objects.OpticalRectangle;
import optics.objects.Refract;
import optics.objects.Wall;

import java.nio.ByteBuffer;

import static application.Storage.opticsTabController;
import static application.Storage.rayTabController;

public class Deserialize {
  public static Serializable deserialize(byte[] object, Pane parent) {
    ByteBuffer buffer = ByteBuffer.wrap(object);
    switch (buffer.getChar()) {
      case 'm': {
        Mirror m = new Mirror(0, 0, 0, 0, parent, 0);
        m.deserialize(object);
        return m;
      }
      case 'w': {
        Wall w = new Wall(0, 0, 0, 0, parent, 0);
        w.deserialize(object);
        return w;
      }
      case 'e': {
        Refract re = new Refract(0, 0, 0, 0, parent, 0, 1);
        re.deserialize(object);
        return re;
      }
      case 'r': {
        Ray r = new Ray(new PreciseLine(new Line()), parent);
        r.deserialize(object);
        return r;
      }
      default:
        return null;
    }
  }

  public static void deserializeAndAdd(byte[] object, Pane parent) {
    Serializable serializable = deserialize(object, parent);
    if (serializable instanceof Ray) {
      rayTabController.createRay((Ray) serializable);
    } else if (serializable instanceof OpticalRectangle) {
      opticsTabController.addObject((OpticalRectangle) serializable, parent);
    }
  }
}

File: src/main/java/serialize/Deserialize.kt
package serialize

import application.Storage
import devices.Device
import devices.Host
import devices.Router
import javafx.scene.layout.Pane

fun deserialize(deviceStrings: List<String>, parent: Pane): List<Device> {
    val devices = deviceStrings.map {
        when (it[0]) {
            'h' -> Host.deserialize(it, parent)
            'r' -> Router.deserialize(it, parent)
            else -> throw Error("Invalid device type")
        }
    }

    deviceStrings.forEach { deviceString ->
        if (deviceString[0] == 'r') {
            val parts = deviceString.split("|")
            val router = devices.find { it.id == parts[1].toInt() } as Router
            val connectedIDs = parts.subList(6, parts.size)
            connectedIDs.filter { it.isNotEmpty() }.forEach { id ->
                val device2 = devices.find { it.id == id.toInt() }!!
                router.addConnection(device2)
            }
        }
    }

    return devices
}
